keps/79-hierarchical-cohorts/README.md

```diff
@@ -0,0 +1,220 @@
+    // share with ClusterQueues within the same group of Cohorts/ClusterQueues.
+    // Each resource group defines the list of resources and a list of flavors
+    // that provide quotas for these resources.
+    // Each resource and each flavor can only form part of one resource group.
+    // resourceGroups can be up to 16.
+    //
+    // BorrowingLimit specifies how much ClusterQueues under this Cohort can borrow
+    // from ClusterQueues/Cohorts that are NOT under this Cohort.
+    //
+    // LendingLimit specifies how much ClusterQueues that are NOT under this Cohort
+    // can borrow from the ClusterQueues/Cohorts that are under this Cohort.
+    // If any of the Limits is not specified it means that there is no limit
+    // and ClusterQueues can borrow/lend as much as they want/have.
+    // 
+    // +listType=atomic
+    // +kubebuilder:validation:MaxItems=16
+    ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`
+}
+
+
+type CohortStatus struct {
+    // conditions hold the latest available observations of the Conditions
+    // current state.
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    // +patchStrategy=merge
+    // +patchMergeKey=type
+    Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+    // Additional stats may be added in the future, like the number 
+    // of admitted workloads, their usage etc, based on the user feedback.
+}
+```
+While admitting the workload Kueue will be checking the whole Cohort subtree whether the
+correct balances are kept. To be more precise what it means
+let's define 2 functions: T(x,r) and D(x,r), that take either ClusterQueue x 
+or Cohort x and resource r (from a specific resource flavor). 
+
+* `T(x, r)` - returns the amount of resource r that is provided at the level of x from ClusterQueues
+and Cohorts that are either x or children of x (possibly indirect). In other words, how much of resource r can come from 
+the subtree. The value may be negative, what means that the subtree is borrowing from the outside of the subtree (the rest 
+of the hierarchy)
+
+* `D(x,r)` - returns the amount of resource r that is provided at the level of x from outside of the subtree. The value may
+be negative, what means that the rest of the hierarchy borrows from the subtree. 
+
+
+Obviously, with the correct admission process, for any x and r, $$T(x,r) + D(x,r)  \geq 0$$
+Otherwise there would be an unbalanced debt at level x. If there is no debt at any level then the admission is correct.
+
+New workload can be admitted to a ClusterQueue if the above conditions stays true at all elements of the hierarchy.
+Checking it only for the ClusterQueue may not be enough as changes related to borrowing may hit limits elsewhere.
+
+`T(x,r)` and `D(x,r)` can be relatively easily calculated while traversing the Cohort tree.
+
+* `T(x,r)` when x is a ClusterQueue:
+$$T(x,r) = quota(x,r) - usage(x,r)$$
```

Comment:
I'm not sure if these formulas are "definitions" or "corollaries" of some other assumptions. Please make it clear is the relationship between the formulas and the statement: "T(x,r) + D(x,r)  >= 0".

Reply:
After some consideration i came to conclusion that we can get rid of D(x,r) in favor of just borrowingLimit(x,r). I pivoted the description there a bit as well. PTAL.

Reply:
Thanks, it reads much better now. I see that the formulas are definitions, and in this section we argue that verifying the condition `T(x,r) + borrowingLimit(x,r) >= 0` at all levels is enough to check the correctness of the admission.

One little puzzling thing is that the definition of `D(x,r)` was different than `borrowingLimit(x,r)`, so it seems the condition actually changed during the re-write. I assume the current version is correct. I will open another dedicated comment if I have some doubts, but I consider the "structural" comment addressed.

Reply:
`D(x,r)` was in fact borrowing limit trimmed to what the rest of the tree can provide. However, we check the provisioning at the level above, so there is no need for extra trimming. It doesn't provide any more information.